### ISSUE TITLE
Post request to download files end with unexpected end of stream exception

### ISSUE THREAD

**Issue by uouo690** [781flyingdutchman/background_downloader#317]
I am trying to download files from private api using post request, where the token send in the body.

An error message indicating that, "unexpected end of stream" 

**Logs**
```
I/BackgroundDownloader( 8120): Enqueuing task with id 2968895671
I/flutter ( 8120): Status: TaskStatus.enqueued
I/TaskWorker( 8120): Starting task with taskId 2968895671
I/flutter ( 8120): Status: TaskStatus.running
I/flutter ( 8120): Progress: 0.0%
D/TaskWorker( 8120): TaskId 2968895671 contentLength undetermined
I/TaskWorker( 8120): Exception for taskId 2968895671: kotlinx.coroutines.JobCancellationException: DispatchedCoroutine is cancelling; job=DispatchedCoroutine{Cancelling}@b08ba2b
W/TaskWorker( 8120): Error for taskId 2968895671: unexpected end of stream
W/TaskWorker( 8120): java.net.ProtocolException: unexpected end of stream
W/TaskWorker( 8120): 	at com.android.okhttp.internal.http.Http1xStream$FixedLengthSource.read(Http1xStream.java:396)
W/TaskWorker( 8120): 	at com.android.okhttp.okio.RealBufferedSource.read(RealBufferedSource.java:50)
W/TaskWorker( 8120): 	at com.android.okhttp.okio.RealBufferedSource.exhausted(RealBufferedSource.java:60)
W/TaskWorker( 8120): 	at com.android.okhttp.okio.GzipSource.read(GzipSource.java:99)
W/TaskWorker( 8120): 	at com.android.okhttp.okio.RealBufferedSource$1.read(RealBufferedSource.java:371)
W/TaskWorker( 8120): 	at java.io.BufferedInputStream.read1(BufferedInputStream.java:286)
W/TaskWorker( 8120): 	at java.io.BufferedInputStream.read(BufferedInputStream.java:347)
W/TaskWorker( 8120): 	at com.bbflight.background_downloader.TaskWorker$transferBytes$2$1.invokeSuspend(TaskWorker.kt:588)
W/TaskWorker( 8120): 	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
W/TaskWorker( 8120): 	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
W/TaskWorker( 8120): 	at kotlinx.coroutines.internal.LimitedDispatcher$Worker.run(LimitedDispatcher.kt:115)
W/TaskWorker( 8120): 	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:100)
W/TaskWorker( 8120): 	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:584)
W/TaskWorker( 8120): 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:793)
W/TaskWorker( 8120): 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:697)
W/TaskWorker( 8120): 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:684)
I/flutter ( 8120): Status: TaskStatus.waitingToRetry
I/flutter ( 8120): Progress: -400.0%
I/WM-WorkerWrapper( 8120): Work [ id=edde6fa0-ff48-4abc-b4a9-3973f0fdbfd1, tags={ com.bbflight.background_downloader.DownloadTaskWorker, BackgroundDownloader, taskId=2968895671, group=default } ] was cancelled
I/WM-WorkerWrapper( 8120): java.util.concurrent.CancellationException: Task was cancelled.
I/WM-WorkerWrapper( 8120): 	at androidx.work.impl.utils.futures.AbstractFuture.cancellationExceptionWithCause(AbstractFuture.java:1183)
I/WM-WorkerWrapper( 8120): 	at androidx.work.impl.utils.futures.AbstractFuture.getDoneValue(AbstractFuture.java:513)
I/WM-WorkerWrapper( 8120): 	at androidx.work.impl.utils.futures.AbstractFuture.get(AbstractFuture.java:474)
I/WM-WorkerWrapper( 8120): 	at androidx.work.impl.WorkerWrapper$2.run(WorkerWrapper.java:316)
I/WM-WorkerWrapper( 8120): 	at androidx.work.impl.utils.SerialExecutorImpl$Task.run(SerialExecutorImpl.java:96)
I/WM-WorkerWrapper( 8120): 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
I/WM-WorkerWrapper( 8120): 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
I/WM-WorkerWrapper( 8120): 	at java.lang.Thread.run(Thread.java:764)
I/BackgroundDownloader( 8120): Enqueuing task with id 2968895671
I/flutter ( 8120): Status: TaskStatus.enqueued
I/TaskWorker( 8120): Starting task with taskId 2968895671
I/flutter ( 8120): Status: TaskStatus.running
I/flutter ( 8120): Progress: 0.0%
D/TaskWorker( 8120): TaskId 2968895671 contentLength undetermined
I/TaskWorker( 8120): Exception for taskId 2968895671: kotlinx.coroutines.JobCancellationException: DispatchedCoroutine is cancelling; job=DispatchedCoroutine{Cancelling}@72462f6
W/TaskWorker( 8120): Error for taskId 2968895671: unexpected end of stream
W/TaskWorker( 8120): java.net.ProtocolException: unexpected end of stream
W/TaskWorker( 8120): 	at com.android.okhttp.internal.http.Http1xStream$FixedLengthSource.read(Http1xStream.java:396)
W/TaskWorker( 8120): 	at com.android.okhttp.okio.RealBufferedSource.read(RealBufferedSource.java:50)
W/TaskWorker( 8120): 	at com.android.okhttp.okio.RealBufferedSource.exhausted(RealBufferedSource.java:60)
W/TaskWorker( 8120): 	at com.android.okhttp.okio.GzipSource.read(GzipSource.java:99)
W/TaskWorker( 8120): 	at com.android.okhttp.okio.RealBufferedSource$1.read(RealBufferedSource.java:371)
W/TaskWorker( 8120): 	at java.io.BufferedInputStream.read1(BufferedInputStream.java:286)
W/TaskWorker( 8120): 	at java.io.BufferedInputStream.read(BufferedInputStream.java:347)
W/TaskWorker( 8120): 	at com.bbflight.background_downloader.TaskWorker$transferBytes$2$1.invokeSuspend(TaskWorker.kt:588)
W/TaskWorker( 8120): 	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
W/TaskWorker( 8120): 	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
W/TaskWorker( 8120): 	at kotlinx.coroutines.internal.LimitedDispatcher$Worker.run(LimitedDispatcher.kt:115)
W/TaskWorker( 8120): 	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:100)
W/TaskWorker( 8120): 	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:584)
W/TaskWorker( 8120): 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:793)
W/TaskWorker( 8120): 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:697)
W/TaskWorker( 8120): 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:684)
I/flutter ( 8120): Progress: -100.0%
I/flutter ( 8120): Status: TaskStatus.failed
I/flutter ( 8120): Download not successful
I/WM-WorkerWrapper( 8120): Work [ id=eb4ff012-a359-470c-afd0-1df576dc93e1, tags={ com.bbflight.background_downloader.DownloadTaskWorker, BackgroundDownloader, taskId=2968895671, group=default } ] was cancelled
I/WM-WorkerWrapper( 8120): java.util.concurrent.CancellationException: Task was cancelled.
I/WM-WorkerWrapper( 8120): 	at androidx.work.impl.utils.futures.AbstractFuture.cancellationExceptionWithCause(AbstractFuture.java:1183)
I/WM-WorkerWrapper( 8120): 	at androidx.work.impl.utils.futures.AbstractFuture.getDoneValue(AbstractFuture.java:513)
I/WM-WorkerWrapper( 8120): 	at androidx.work.impl.utils.futures.AbstractFuture.get(AbstractFuture.java:474)
I/WM-WorkerWrapper( 8120): 	at androidx.work.impl.WorkerWrapper$2.run(WorkerWrapper.java:316)
I/WM-WorkerWrapper( 8120): 	at androidx.work.impl.utils.SerialExecutorImpl$Task.run(SerialExecutorImpl.java:96)
I/WM-WorkerWrapper( 8120): 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
I/WM-WorkerWrapper( 8120): 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
I/WM-WorkerWrapper( 8120): 	at java.lang.Thread.run(Thread.java:764)

```

**Code**
```
final task = bacD.DownloadTask(
        post: {token: "******"},
        headers: {
          'Accept': '*/*',
          'content-type': 'application/json',
        },
        url: '*************',
        // urlQueryParameters: {'q': 'pizza'},
        filename: 'results.pdf',
        directory: 'my_sub_directory',
        updates:
            Updates.statusAndProgress, // request status and progress updates
        requiresWiFi: true,
        retries: 1,
        metaData: 'data for me');
        
        final result = await bacD.FileDownloader().download(task,
        onProgress: (progress) => print('Progress: ${progress * 100}%'),
        onStatus: (status) => print('Status: $status'));
```

I can download those files using chrome browser, but I can't get it to work in my flutter app.
Thanks in advance.

**Comment by 781flyingdutchman** [comment]
The `post` argument of the constructor needs to be a String, so you need to `jsonEncode` the map you provide

**Comment by 781flyingdutchman** [comment]
Actually just published V8.5.2 which does take a Map as the post argument and jsonEncodes it for you (provided it is simple and does not contain complex objects that require custom encoding)

**Comment by daviddomkar** [comment]
I had this issue as well but maybe for a different reason, created a PR which fixes this: https://github.com/781flyingdutchman/background_downloader/pull/533
